google/detectors/credentials/generic_weak_credential_detector/src/main/java/com/google/tsunami/plugins/detectors/credentials/genericweakcredentialdetector/testers/grafana/GrafanaCredentialTester.java

```diff
@@ -0,0 +1,196 @@
+
+/** Credential tester specifically for grafana. */
+public final class GrafanaCredentialTester extends CredentialTester {
+  private static final GoogleLogger logger = GoogleLogger.forEnclosingClass();
+  private final HttpClient httpClient;
+
+  private static final String GRAFANA_SERVICE = "grafana";
+  private static final String GRAFANA_PAGE_TITLE = "grafana";
+
+  @Inject
+  GrafanaCredentialTester(HttpClient httpClient) {
+    this.httpClient = checkNotNull(httpClient);
+  }
+
+  @Override
+  public String name() {
+    return "GrafanaCredentialTester";
+  }
+
+  @Override
+  public String description() {
+    return "Grafana credential tester.";
+  }
+
+  private static String buildTargetUrl(NetworkService networkService, String path) {
+    StringBuilder targetUrlBuilder = new StringBuilder();
+
+    if (NetworkServiceUtils.isWebService(networkService)) {
+      targetUrlBuilder.append(NetworkServiceUtils.buildWebApplicationRootUrl(networkService));
+
+    } else {
+      // Default to HTTP protocol when the scanner cannot identify the actual service.
+      targetUrlBuilder
+          .append("http://")
+          .append(NetworkEndpointUtils.toUriAuthority(networkService.getNetworkEndpoint()))
+          .append("/");
+    }
+    targetUrlBuilder.append(path);
+    return targetUrlBuilder.toString();
+  }
+
+  /**
+   * Determines if this tester can accept the {@link NetworkService} based on the name of the
+   * service or a custom fingerprint. The fingerprint is necessary since nmap doesn't recognize a
+   * grafana instance correctly.
+   *
+   * @param networkService the network service passed by tsunami
+   * @return true if a grafana instance is recognized
+   */
+  @Override
+  public boolean canAccept(NetworkService networkService) {
+
+    boolean canAcceptByNmapReport =
+        NetworkServiceUtils.getWebServiceName(networkService).equals(GRAFANA_SERVICE);
+    boolean canAcceptByCustomFingerprint = false;
+
+    var url = buildTargetUrl(networkService, "");
+    try {
+      logger.atInfo().log("probing Grafana home - custom fingerprint phase");
+
+      HttpResponse response = httpClient.send(get(url).withEmptyHeaders().build());
+      canAcceptByCustomFingerprint =
+          response.status().isSuccess()
+              && response
+                  .bodyString()
+                  .map(GrafanaCredentialTester::bodyContainsGrafanaElements)
+                  .orElse(false);
+    } catch (IOException e) {
+      logger.atWarning().withCause(e).log("Unable to query '%s'.", url);
+      return false;
+    }
+
+    return canAcceptByNmapReport || canAcceptByCustomFingerprint;
+  }
+
+  // Checks if the response body contains elements of a grafana page - custom fingerprinting phase
+  private static boolean bodyContainsGrafanaElements(String responseBody) {
+    Document doc = Jsoup.parse(responseBody);
+    String title = doc.title();
+
+    if (title.toLowerCase().contains(GRAFANA_PAGE_TITLE)) {
```

Comment:
This assumption might be insufficient. Any page that contains "grafana" in the title will match. For example "How to update your Grafana", which will be a static page. Catching more things in the fingerprinting phase might generally be good, but given that brute-forcing is time consuming, I would suggest finding a stronger assumption.

Reply:
I improved the fingerprint by checking more elements in the page and the `/api/health` unauthenticated endpoint

See #5482574 and #884695e